server/game/cardsnapshot.js

```diff
@@ -0,0 +1,68 @@
+const _ = require('underscore');
```

Comment:
Why not just copy the card itself? Seems you’re taking most of the core properties anyway

Reply:
If you mean create another `DrawCard` object with the same properties - I explicitly separated it into its own class to get rid of the mutator methods. It would be best if the snapshot was a read-only object and this seemed like the best way. I'd rather code crash explicitly if you try to call, say, `modifyStrength` on a snapshot than it introduce subtle bugs by silently succeeding at modifying a card that's not real.

Reply:
The flip side is that now you have two places to maintain state and if a new card needs to react to state not provided already then it'll need adding into this.  Or if something core needs changing (admittedly unlikely) it'll need changing in both places.  I take the point about mutation, but if you make changes to an object representing historical state and not current state then you should expect it not to work as you intended :man_shrugging:

Reply:
Agreed that it's not ideal that they parallel each other to the degree that they do. I think the larger problem with that is the logic duplication of checking factions, traits, etc than anything else. That could be solved by extracting those things to their own classes / static methods and using them both places.

Changes to either place at this point would require for an entirely new game element (like cost, strength, trait, etc) be introduced. Which as you say, is unlikely. Though it does look like I missed at least one thing (icons) - I was trying to get them all in one go so we don't have to worry about "a new card needs X"

On the mutation, obviously if someone does the following then they should not expect it to work:
```javascript
event.cardSnapshot.modifyStrength(1)
```

I'm more worried about subtle cases where a card gets stored off or passed around a bit and it isn't clear whether it's a card or a snapshot. If this were a statically typed language, I'd use a common interface between the two for read-only ops. As it is, the safest option seems to be crash horribly or accept that sometime somewhere someone is going to pass an object around, modify it, and it isn't obvious what went wrong.
```javascript
// This is contrived as it assumes all methods are on the same class. Could 
// easily be passed from object to object and attempt to modify the snapshot 
// a long ways away on a different class from the original call
this.doSomething(event.cardSnapshot)

doSomething(c) {
    this.doSomethingElse(c);
}

doSomethingElse(c) {
    c.modifyStrength(1);
}
```

Reply:
I'd say that is equally unlikely given that you're mostly talking about event handlers here, which by their nature should be pretty concise.  Things that make changes to card state should really go in effects in most cases.  It almost feels like this should inherit from one of the card base objects, and if that still contains too much perhaps we need a new base object?